rfc/rfc021-vp_token-grant-type.md

```diff
@@ -0,0 +1,189 @@
+
+## 3. HTTP Parameter Bindings for Transporting Assertions
+
+The OAuth Assertion Framework [RFC7521] defines generic HTTP parameters for transporting assertions (a.k.a. security tokens) during interactions with a token endpoint. 
+This section defines specific parameters and treatments of those parameters for use with VP Bearer Tokens as Authorization Grants.
+
+To use a VP as an authorization grant, the client uses an access token request as defined in Section 4 of the OAuth Assertion Framework [RFC7521] with the following specific parameter values and encodings.
+
+The value of the "grant_type" is "vp_token-bearer".
+
+The value of the "assertion" parameter MUST contain a Verifiable Presentation. The Verifiable Presentation MUST be encoded as either:
+
+* The Verifiable Presentation is encoded as JSON using the application/x-www-form-urlencoded content type.
+* The Verifiable Presentation is encoded as JWT according to section §6.3.1 of the Verifiable Credentials Data Model 1.1 [VC-DATA-MODEL].
+
+The "scope" parameter MUST be used, as defined in the OAuth Assertion Framework [RFC7521], to indicate the requested scope. 
+The scope parameter determines the set of credentials the Authorization Server expects.
+
+The "presentation_submission" parameter MUST be used, as defined in Presentation Exchange [PE], to indicate how the VP matches the requested Verifiable Credentials.
+The Presentation Submission MUST be encoded using the application/x-www-form-urlencoded content type.
+
+The following example demonstrates an access token request with a VP as an authorization grant (with extra line breaks for display purposes only):
+
+     POST /token.oauth2 HTTP/1.1
+     Host: as.example.com
+     Content-Type: application/x-www-form-urlencoded
+
+     grant_type=vp_token-bearer
+     &assertion=eyJhbGciOiJFUzI1NiIsImtpZCI6IjE2In0.
+     eyJpc3Mi[...omitted for brevity...].
+     J9l-ZhwP[...omitted for brevity...]
+     &presentation_submission= {..}
+     &scope=a_scope
+
+## 3.1 Assertion Encoding
+
+The assertion parameter can either be encoded in JSON or JWT.
+The Authorization Server determines the encoding by adding the correct parameters to the Authorization Server metadata [RFC8414]].
+The following parameter MUST be added:
+
+* `vp_formats`: An object defining the formats and proof types of Verifiable Presentations and Verifiable Credentials that a Verifier supports as stated by §9 of the OpenID for Verifiable Presentations specification [OIDC4VP].
+
+When processing the assertion the Authorization Server can detect the used format by checking the initial character.
+A JSON encoded assertion starts with a '{' character or '%7B' when urlencoded.
```

Comment:
Presentation submission should say if its JWT or JSON?

Reply:
Even so, format still has to be detected. Maybe better to remove comment entirely?

Reply:
A submission only details the types of the VCs not the VP

Reply:
From [OpenID4VP](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-response-parameters)
> vp_token: REQUIRED. JSON String or JSON object that MUST contain a single Verifiable Presentation or an array of JSON Strings and JSON objects each of them containing a Verifiable Presentations...

So the `vp_token` consists of 1 or more VPs, and the credentials are always nested in the VPs. (OpenID4VP allows nesting 1 level deep exactly for this reason.)

See example from the link
```
{
    "id": "Presentation example 1",
    "definition_id": "Example with selective disclosure",
    "descriptor_map": [
        {
            "id": "ID card with constraints",
            "format": "ldp_vp",
            "path": "$",
            "path_nested": {
                "format": "ldp_vc",
                "path": "$.verifiableCredential[0]"
            }
        }
    ]
}
```

This specifies the format of the VP

Reply:
After further assessment I now believe PEX/JSONPath also allows the above to be expressed as
```
{
    "id": "Presentation example 2",
    "definition_id": "Example with multiple VPs",
    "descriptor_map": [
        {
            "id": "ID Card with constraints",
            "format": "ldp_vc",
            "path": "$[0].verifiableCredential[0]"
        },
        {
            "id": "Ontario Health Insurance Plan",
            "format": "jwt_vc_json",
            "path": "$[1].vp.verifiableCredential[0]"
        }
    ]
}
```
In this case the format of the VP is not known...

This stuff is too complex...

Reply:
encoding line has been removed, so conversation is no longer valid or no longer on topic.